File: programmers/src/큰수만들기.kt
import kotlin.math.max

class Solution {
    fun solution(number: String, k: Int): String {
        var answer = StringBuilder()
        var index = 0
        var maxValue = 0
        // number의 길이에서 k 만큼을 뺀 길이의 수가 필요하므로 그 만큼 반복하게 된다.
        for (i in 0 until number.length - k) {
            maxValue = 0
            // 끝 범위 설정 : i + k + 1 - 1 = i + k
            for (j in index..i + k) {
               if (maxValue < number[j].digitToInt()){
                   index = j+1
                   maxValue = number[j].digitToInt()
               }
            }
            answer.append(maxValue)
        }
        return answer.toString()
    }
}

// 완전 탐색풀이 -> 메모리 부족으로 ㅋ통과 못함
// 인덱스만큼 뒷자리와 맨 앞자리 비교하기
// 맨뒷값이 추가하려는 값보다 큰지

//0 0 0 0 0
//1 2 3 4
//
//fun main() {
//    val sol = Solution()
//    sol.solution("1231234", 3)
//}
//
//class Solution {
//
//    var dfsList = mutableListOf<MutableList<Int>>()
//    fun solution(number: String, k: Int): String {
//        var answer: Long = Int.MIN_VALUE.toLong()
//        dfs(number.length, k, mutableListOf<Int>(-1))
//        dfsList.forEach {
//            val newString = StringBuilder(number)
//            it.forEach {
//                newString.setCharAt(it, ' ')
//            }
//            answer = max(answer,newString.toString().replace(" ", "").toLong())
//        }
//        return answer.toString()
//    }
//
//    private fun dfs(number: Int, k: Int, list: MutableList<Int>) {
//        if (list.size - 1 == k) {
//            list.removeLast()
//            dfsList.add(list)
//        } else if (list[0] > number - 1) return
//        else {
//            for (i in list[0] + 1 until number) {
//                val newList = mutableListOf<Int>()
//                newList.addAll(list)
//                newList.add(0, i)
//                dfs(number, k, newList)
//            }
//        }
//    }
//}

File: Programmers/그리디/큰 수 만들기.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

string solution(string number, int k) {
	string answer = "";

	// 구하려는 수의 자릿수
	int digitCount = number.length() - k;
	// cout<< digitCount << "\n";

	int findInx = -1;
	for (int i = digitCount; i >= 1; i--) {

		char max = '0';
		int inx = -1;

		for (int j = number.length() - i; j > findInx; j--) {

			char current = number[j];
			if (current >= max) {
				max = current;
				inx = j;
			}

		}

		findInx = inx; //현재 자릿수로 뽑힌 수의 위치, 다음 자릿수는 이 위치 전 까지 탐색 가능
		answer += max;

		// cout << "findInx : " << findInx << "\n";
	}


	return answer;
}


// 뒤에서 부터 만드려는 수의 각 자릿수 별로 최대값을 찾는 방식